File: 알고리즘 with 동빈나/정렬 알고리즘/Python1/Ex_quickSort.py
# 퀵 정렬 평균 시간복잡도: O(NlongN), 최악: O(N^2)
array = [7, 5, 9, 0, 3, 1, 6, 2, 4, 8]

# 퀵 정렬 함수(배열, 시작 인덱스, 끝 인덱스)
def quick_sort(array, start, end):
    if start >= end:  # 원소가 1개인 경우 종료
        return

    pivot = start  # 피벗(= 기준값)은 첫 번째 원소
    left = start + 1
    right = end

    # left와 right가 가르키는 값이 서로 교차하지 않으면: 참
    while left <= right:
        # 피벗보다 큰 데이터를 찾을 때까지 반복
        while (left <= end and array[left] <= array[pivot]):
            left += 1

        # 피벗보다 작은 데이터를 찾을 때까지 반복
        while(right > start and array[right] >= array[pivot]):
            right -= 1

        # 엇갈린 경우 left가 가리킨 데이터와 pivot을 swap
        if left > right:
            array[right], array[pivot] = array[pivot], array[right]
        else: # left와 right가 가리키는 데이터 swap
            array[left], array[right] = array[right], array[left]

    # 분할 이후 pivot을 기준으로 왼쪽과 오른쪽 각각 정렬 수행 
    quick_sort(array, start, right - 1)  # array[right]: pivot
    quick_sort(array, right + 1, end)  # array[right]: pivot


quick_sort(array, 0, len(array)-1)
print(array)

File: 알고리즘 with 동빈나/정렬 알고리즘/Cpp1/Ex_quickSort.cpp
#ifndef BITS_STD_H
#define BITS_STD_H
#include <bits/stdc++.h>
using namespace std;

#endif // !BITS_STD_H
static void quickSort(int arr[], int start, int end);

int main_ex_quick_sort()
{
	const int len = 10;
	int arr[len] = { 7, 5, 9, 0, 3, 1, 6, 2, 4, 8 };

	quickSort(arr, 0, len - 1);

	for (int i = 0; i < len; ++i)
	{
		cout << arr[i] << ' ';
	}
	cout << endl;

	return EXIT_SUCCESS;
}

static void quickSort(int arr[], int start, int end)
{
	if (start >= end)
		return;

	int pivot_i = start;
	int left_i = start + 1;
	int right_i = end;

	while (left_i <= right_i)
	{

		while (arr[left_i] < arr[pivot_i])
			left_i += 1;

		while (arr[right_i] > arr[pivot_i])
			right_i -= 1;

		if (left_i <= right_i)
		{
			swap(arr[left_i], arr[right_i]);
		}
		else
		{
			swap(arr[pivot_i], arr[right_i]);
		}
	}

	quickSort(arr, start, right_i - 1);
	quickSort(arr, right_i + 1, end);
}